src/lib/Network.java

```diff
@@ -24,79 +30,63 @@ public static byte[] compress(byte[] data) throws IOException {
-            if (GZip) dOS.write("Content-Encoding: gzip\r\n".getBytes());
-            temp.append("Content-Type: ");
-            temp.append(bs);
-            if (bs.equals("text/html")) temp.append(";charset=UTF-8\r\n");
-            else temp.append("\r\n");
-            dOS.write(temp.toString().getBytes());
-            temp.setLength(0);
-            if (!isFile) {
-                temp.append("Content-Length: ");
-                temp.append(ResponseData.length);
+        });
+        temp.append("content-type: ");
+        temp.append(httpResponse.getHttpContentType());
+        temp.append("\r\n");
+
+        try {
+            if (gzip) {
+                byte[] body = compress(httpResponse.getBody().getBytes());
+                temp.append("content-encoding: gzip\r\n");
+                temp.append("content-length: ");
+                temp.append(body.length);
+                temp.append("\r\n\r\n");
+                outputStream.write(temp.toString().getBytes());
+                outputStream.write(body);
+            } else {
+                temp.append("content-length: ");
+                temp.append(httpResponse.getBody().length());
+                temp.append("\r\n\r\n");
+                temp.append(httpResponse.getBody());
+                outputStream.write(temp.toString().getBytes());
             }
-            temp.append("\r\n\r\n");
-            dOS.write(temp.toString().getBytes());
-            if (isFile && !GZip) {
-                FileInputStream is = new FileInputStream(new String(ResponseData));
-                BufferedInputStream bis = new BufferedInputStream(is, 1024);
-                byte[] b;
-                while ((b = bis.readNBytes(1024)).length != 0) dOS.write(b);
-            } else dOS.write(ResponseData);
-            dOS.flush();
-            dOS.close();
+            outputStream.close();
         } catch (Exception e) {
             log.e(e, Network.class.getName(), "write");
         }
     }
 
     /**
-     * Reads from socket into ArrayList
+     * Reads a bare request from a socket
      *
-     * @param MAX_REQ_SIZE the maximum kbytes to read
+     * @param maxReqSize the maximum bytes to read
+     * @return String representing the request
      */
-    public static ByteArrayOutputStream read(BufferedInputStream dIS, int MAX_REQ_SIZE) {
-        MAX_REQ_SIZE = MAX_REQ_SIZE * 1000;
-        ByteArrayOutputStream Reply = new ByteArrayOutputStream(1024);
-        int counter = 0;
-        try {
-            do {
-                if (counter < MAX_REQ_SIZE) {
-                    Reply.write(dIS.read());
-                    counter++;
-                } else break;
-            } while (dIS.available() > 0);
-        } catch (Exception ignored) {
+    public static String read(BufferedInputStream inputStream, int maxReqSize) throws IOException {
+        byte[] buffer = new byte[maxReqSize];
+        if (inputStream.read(buffer, 0, maxReqSize) == maxReqSize) {
```

Comment:
The read function might not be able to report an accurate number. Moreover, if the remaining reqSize is x and the request is larger, we should read the x amount only. TLS Records handling is a bit tricky.

Reply:
The `read` function is guaranteed to read until

- The specified number of bytes have been read,
- The read method of the underlying stream returns -1, indicating end-of-file, or
- The available method of the underlying stream returns zero, indicating that further input requests would block. 

if it reads the maximum amount of bytes this means that the request is bigger than the configured maximum and the request should be dropped.

What needs to be improved is the exception handling, but this is a work for another day.

Reply:
The available method is not reliable enough in TLS communication. [Check this discussion I made a year ago](https://coderanch.com/t/747813/java/SSL-TLS-Record-reading-java).

![image](https://user-images.githubusercontent.com/48622891/224838424-31412f89-aa7f-454d-914d-ffbc5e0ac558.png)

Reply:
We can sort it out if we experimented sending +16kb post data over tls and see how it is going to read

Reply:
I hope the new commit fixes this problem